# Title: Segmenting and Clustering Neighborhoods in Toronto, Canada 

Before we start with the assignment content, let's download all the dependencies that we will need.

In [183]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## Download and Explore Dataset

Neighborhood has a total of 11 boroughs and 208 neighborhoods, In order to segment the neighborhoods and explore them, we will essentially need a dataset that contains the 11 boroughs and the neighborhoods that exist in each borough.

#### Download the data and Tranform it into a *pandas* dataframe

In [184]:
data = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")

Let's take a quick look at the data.

In [185]:
data[0].head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

Replace the column names as per requirement of the assignment.

In [186]:
data[0].rename(columns={'Postcode':'PostalCode'},inplace=True)
data[0].rename(columns={'Neighbourhood':'Neighborhood'},inplace=True)
data[0].head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

As Per the requirement of the assignment we will only process the cells that have an assigned borough and ignore or drop those which have "Not assigned" boroughs.

In [187]:
indexNames = data[0][ data[0]['Borough'] == 'Not assigned' ].index
data[0].drop(indexNames , inplace=True)
data[0].reset_index(inplace=True)
cleaned_data=data[0].drop('index', axis=1)

Lets look at the data after cleaning boroughs that shows "Not assigned".

In [188]:
cleaned_data.head()

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

As Per the requirement of the assignment if a cell has a borough but a "Not assigned" neighborhood, then the neighborhood will be the same as the borough.

In [189]:
cleaned_data.Neighborhood[cleaned_data.Neighborhood== 'Not assigned'] = cleaned_data.Borough

Lets look at the Data

In [190]:
cleaned_data.head()

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

Lets Check the number of records available in the dataframe.

In [191]:
cleaned_data.shape

(211, 3)

So, there are 211 records in the data frame after removing the "Not Assigned" neighborhood records. 
Now, lets see how many records we will have when we combine more than one neighborhood in one postal code area one row with the neighborhoods separated with a comma.

In [192]:
agg_data = cleaned_data.groupby(['PostalCode','Borough'], as_index=False).agg({'Neighborhood' : lambda x: ', '.join(x)})

In [193]:
agg_data.shape

(103, 3)

So after the complete cleanup we have 103 records in the data frame.

In [194]:
agg_data

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood  
0                                       Rouge, Malvern  
1               Highland Creek, Rouge Hill, Port Union  
2                    Guildwood, Morningside, West Hill  
3                              

In order to get the geographical coordinates of the neighborhoods using the Geocoder package, We will use the csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data. We will load the csv to IBM DB2 and connect the notebook to the DB2 database table.

Lets include the dependancies of IBM DB2 instance.

In [195]:
import ibm_db

Connect to IBM DB2 using service credentials.

In [196]:
dsn_driver = "{IBM DB2 ODBC DRIVER}"
dsn_database = "BLUDB"            
dsn_hostname = "dashdb-txn-sbox-yp-dal09-04.services.dal.bluemix.net"           
dsn_port = "50000"                  
dsn_protocol = "TCPIP"            
dsn_uid = "spw42978"                
dsn_pwd = "p@2mhzhsh4mvqqkz"     

In [197]:
dsn = (
    "DRIVER={0};"
    "DATABASE={1};"
    "HOSTNAME={2};"
    "PORT={3};"
    "PROTOCOL={4};"
    "UID={5};"
    "PWD={6};").format(dsn_driver, dsn_database, dsn_hostname, dsn_port, dsn_protocol, dsn_uid, dsn_pwd)

try:
    conn = ibm_db.connect(dsn, "", "")
    print ("Connected to database: ", dsn_database, "as user: ", dsn_uid, "on host: ", dsn_hostname)

except:
    print ("Unable to connect: ", ibm_db.conn_errormsg() )


Connected to database:  BLUDB as user:  spw42978 on host:  dashdb-txn-sbox-yp-dal09-04.services.dal.bluemix.net


In [198]:
import ibm_db_dbi
#connection for pandas
pconn = ibm_db_dbi.Connection(conn)

In [199]:
#query statement to retrieve all rows in GEOSPATIAL_COORDINATES table
selectQuery = "select * from GEOSPATIAL_COORDINATES"

#retrieve the query results into a pandas dataframe
pdf = pd.read_sql(selectQuery, pconn)


Lets look at the postalcodes data from the csv with lattitude and longitude. 

In [200]:
pdf.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

Now join the pdf data frame (postalcode with lattitude and longitude) and the agg_data data frame that contains the cleaned data.

In [201]:

joining = pd.merge(agg_data, pdf, on='PostalCode', how='inner')
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(joining['Borough'].unique()),
        joining.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


In [202]:
joining

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    G

#### Use geopy library to get the latitude and longitude values of Toronto.

In [203]:
address = 'Toronto, CN'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6425637, -79.3870871832047.


In [205]:
# create map of toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(joining['Latitude'], joining['Longitude'], joining['Borough'], joining['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [206]:
CLIENT_ID = '40GJXZK22WHX4TGOS2PJO0MPHSMA5I05NRIPX450ZIIMGYOI' # your Foursquare ID
CLIENT_SECRET = 'T3FR0XUNMVJAOUSC52NXNX3FN12BFP1OYFHX1Y5EW0BBYGJY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('40GJXZK22WHX4TGOS2PJO0MPHSMA5I05NRIPX450ZIIMGYOI: ' + CLIENT_ID)
print('T3FR0XUNMVJAOUSC52NXNX3FN12BFP1OYFHX1Y5EW0BBYGJY:' + CLIENT_SECRET)

Your credentails:
40GJXZK22WHX4TGOS2PJO0MPHSMA5I05NRIPX450ZIIMGYOI: 40GJXZK22WHX4TGOS2PJO0MPHSMA5I05NRIPX450ZIIMGYOI
T3FR0XUNMVJAOUSC52NXNX3FN12BFP1OYFHX1Y5EW0BBYGJY:T3FR0XUNMVJAOUSC52NXNX3FN12BFP1OYFHX1Y5EW0BBYGJY


#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name for those boroughs that contains name "Toronto"

In [207]:
toronto_data = joining[joining['Borough'].str.contains("Toronto")].reset_index(drop=True)
toronto_data.loc[0, 'Neighborhood']

'The Beaches'

Get the neighborhood's latitude and longitude values.

In [208]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.6763574, -79.2930312.


#### Now, let's get the top 100 venues that are in "The Beaches" within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [209]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=40GJXZK22WHX4TGOS2PJO0MPHSMA5I05NRIPX450ZIIMGYOI&client_secret=T3FR0XUNMVJAOUSC52NXNX3FN12BFP1OYFHX1Y5EW0BBYGJY&v=20180605&ll=43.6763574,-79.2930312&radius=500&limit=100'

Send the GET request and examine the results

In [210]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ca859a7db04f507f70929b6'},
 'response': {'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.6808574045, 'lng': -79.28682091449052},
   'sw': {'lat': 43.6718573955, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4ad4c062f964a52011f820e3',
       'name': 'The Big Carrot Natural Food Market',
       'location': {'address': '125 Southwood Dr',
        'lat': 43.678879,
        'lng': -79.297734,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.678879,
          'lng': -79.297734}],
        'distance': 471,
        'postalCode': 'M4E 0B8',
        'cc': 'C

In [211]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [212]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name         categories        lat        lng
0  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
1                 Grover Pub and Grub                Pub  43.679181 -79.297215
2                           Starbucks        Coffee Shop  43.678798 -79.298045
3                       Upper Beaches       Neighborhood  43.680563 -79.292869

In [213]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


## Explore Neighborhoods in Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [214]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [215]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )


The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

#### Let's check the size of the resulting dataframe

In [216]:
print(toronto_venues.shape)
toronto_venues.head()

(1710, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031  The Big Carrot Natural Food Market       43.678879   
1              -79.293031                 Grover Pub and Grub       43.679181   
2              -79.293031                           Starbucks       43.678798   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  
0       -79.297734  Health Food Store  
1       -79.297215                Pub  
2       -79.298045        Coffee Shop  
3       -79.292869       Neighborhood  
4       -79.351434   Greek Restaurant

Let's check how many venues were returned for each neighborhood

In [217]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            55   
Brockton, Exhibition Place, Parkdale Village                           20   
Business Reply Mail Processing Centre 969 Eastern                      17   
CN Tower, Bathurst Quay, Island airport, Harbou...                     13   
Cabbagetown, St. James Town                                            43   
Central Bay Street                                                     84   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   89   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             37   
Davisville North                                                       11   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     15   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           20   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         36   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              48   
High Park, The Junction South                                          25   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               64   
Moore Park, Summerhill East                                             4   
North Toronto West                                                     19   
Parkdale, Roncesvalles                                                 15   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     39   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        97   
Studio District                                                        37   
The Annex, North Midtown, Yorkville                                    25   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         22   
The Danforth West, Riverdale                                           43   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             55   
Brockton, Exhibition Place, Parkdale Village                            20   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower, Bathurst Quay, Island airport, Harbou...                      13   
Cabbagetown, St. James Town                                             43   
Central Bay Street                                                      84   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    89   
Commerce Court, Victoria Hotel     

#### Let's find out how many unique categories can be curated from all the returned venues

In [218]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 240 uniques categories.


## Analyze Each Neighborhood

In [219]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()


Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Beer Bar  Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  \
0         0           0                   0       0              0          0   
1         0           0                   0       0              0          0   
2         0           0                   0       0              0          0   
3         0           0                   0       0              0          0   
4         0           0                   0       0              0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0                     0               0        0                0   
1         0                     0               0        0                0   
2         0                     0               0        0                0   
3         0                     0               0        0                0   
4         0                     0               0        0                0   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0         0             0              0         0        0     0   
1         0             0              0         0        0     0   
2         0             0              0         0        0     0   
3         0

And let's examine the new dataframe size.

In [220]:
toronto_onehot.shape

(1710, 240)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [221]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.000000   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.011905   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.022472   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.027778   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.020833   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.015625   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.052632   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.000000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.000000   
32                    Stn A PO Boxes 25 The Esplanade     0.000000   
33                                    Studio District     0.027027   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023256   

    Accessories Store  Adult Boutique  Afghan Restaurant   Airport  \
0                0.01        0.000000           0.000000  0.000000   
1                0.00        0.000000           0.000000  0.000000   
2                0.00        0.000000           0.000000  0.000000   
3                0.00        0.000000           0.000000  0.000000   
4                0.00        0.000000           0.000000  0.076923   
5                0.00        0.000000           0.000000  0.000000   
6                0.00        0.000000           0.000000  0.000000   
7                0.00        0.000000           0.000000  0.000000   
8                0.00        0.000000           0.000000  0.000000   
9                0.00        0.011236           0.011236  0.000000   
10               0.00        0.000000           0.000000  0.000000   
11               0.00        0.000000           0.000000  0.000000   
12               0.00        0.000000           0.000000  0.000000   
13               0.00        0.000000           0.000000  0.000000   
14               0.00        0.000000           0.000000  0.000000   
15               0.00        0.000000           0.000000  0.000000   
16               0.00        0.000000           0.000000  0.000000   
17               0.00        0.000000           0.

#### Let's confirm the new size

In [222]:
toronto_grouped.shape

(38, 240)

#### Let's print each neighborhood along with the top 5 most common venues

In [223]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.06
1             Café  0.04
2  Thai Restaurant  0.04
3       Steakhouse  0.04
4              Bar  0.04


----Berczy Park----
          venue  freq
0   Coffee Shop  0.07
1  Cocktail Bar  0.05
2    Restaurant  0.04
3    Steakhouse  0.04
4          Café  0.04


----Brockton, Exhibition Place, Parkdale Village----
                    venue  freq
0                    Café  0.10
1          Breakfast Spot  0.10
2             Coffee Shop  0.10
3           Grocery Store  0.05
4  Furniture / Home Store  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.12
1         Pizza Place  0.06
2    Recording Studio  0.06
3             Brewery  0.06
4                 Spa  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0  Airport Terminal  0.15
1   Airport Service

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [224]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [225]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village        Breakfast Spot   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...      Airport Terminal   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0       Thai Restaurant                  Café            Steakhouse   
1          Cocktail Bar            Restaurant                  Café   
2           Coffee Shop                  Café                   Gym   
3                Garden         Garden Center         Auto Workshop   
4       Airport Service        Airport Lounge          Airport Gate   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                     Bar                 Hotel                   Gym   
1          Farmers Market                Bakery            Steakhouse   
2  Furniture / Home Store    Falafel Restaurant                   Bar   
3        Recording Studio            Skate Park               Brewery   
4        Sculpture Garden                 Plane       Harbor / Marina   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant                Bakery    American Restaurant  
1                   Pub    Seafood Restaurant            Cheese Shop  
2               Stadium         Burrito Place             Restaurant  
3  Fast Food Restaurant                   Spa          Burrito Place  
4    Airport Food Court               Airport          Boat or Ferry

## Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [226]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [227]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                   Pub     Health Food Store   
1 -79.352188               1      Greek Restaurant           Coffee Shop   
2 -79.315572               1        Sandwich Place           Coffee Shop   
3 -79.340923               1                  Café           Coffee Shop   
4 -79.388790               4              Bus Line                  Park   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop           Wings Joint                 Diner   
1        Ice Cream Shop    Italian Restaurant     Indian Restaurant   
2        Ice Cream Shop          Liquor Store    Light Rail Station   
3    Italian Restaurant             Gastropub   American Restaurant   
4           Swim School           Wings Joint    Falafel Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Discount Store               Dog Run      Doner Restaurant   
1                Bakery            Sports Bar   Japanese Restaurant   
2          Burger Joint         Burrito Place  Fast Food Restaurant   
3                Bakery               Brewery                  Bank   
4           Event Space  Ethiopian Restaurant     Electronics Store   

         9th Most Common Venue       10th Most Common Venue  
0                   Donut Shop  Eastern European Restaurant  
1                          Spa                    Juice Bar  
2            Fish & Chips Shop                   Steakhouse  
3                          Bar                  Fish Market  
4  Eastern European Restaurant          Dumpling Restaurant

Finally, let's visualize the resulting clusters

In [229]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters

#### Toronto Business Center Cluster

This cluster is a mix of businesses like restaurants, Bars, pet track and medical center etc, thats why we have named it as a Business center cluster because it is a combination of multiple businesses.

In [230]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0     East Toronto               0                   Pub   
9  Central Toronto               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Health Food Store           Coffee Shop           Wings Joint   
9                   Pub        Medical Center   American Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Diner        Discount Store               Dog Run   
9      Sushi Restaurant            Bagel Shop   Fried Chicken Joint   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0      Doner Restaurant            Donut Shop  Eastern European Restaurant  
9            Sports Bar           Supermarket            Convenience Store

#### Toronto Down Town Cluster

This cluster is mainly showing all the business located in a downtown of toronto. Everything that a downtown can have is listed in this cluster like restaurants, bars, hotels, Markets etc. Thats why we have named this as a toronto down town cluster because the downtown borough has occured multiple times and its the largest cluster among all the 5 clusters. 

In [231]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1       East Toronto               1      Greek Restaurant   
2       East Toronto               1        Sandwich Place   
3       East Toronto               1                  Café   
5    Central Toronto               1           Pizza Place   
6    Central Toronto               1   Sporting Goods Shop   
7    Central Toronto               1        Sandwich Place   
8    Central Toronto               1            Restaurant   
11  Downtown Toronto               1           Coffee Shop   
12  Downtown Toronto               1           Coffee Shop   
13  Downtown Toronto               1           Coffee Shop   
14  Downtown Toronto               1           Coffee Shop   
15  Downtown Toronto               1           Coffee Shop   
16  Downtown Toronto               1           Coffee Shop   
17  Downtown Toronto               1           Coffee Shop   
18  Downtown Toronto               1           Coffee Shop   
19  Downtown Toronto               1           Coffee Shop   
20  Downtown Toronto               1           Coffee Shop   
21  Downtown Toronto               1           Coffee Shop   
24   Central Toronto               1           Coffee Shop   
25  Downtown Toronto               1                  Café   
26  Downtown Toronto               1                  Café   
27  Downtown Toronto               1      Airport Terminal   
28  Downtown Toronto               1           Coffee Shop   
29  Downtown Toronto               1           Coffee Shop   
30  Downtown Toronto               1                  Café   
31      West Toronto               1              Pharmacy   
32      West Toronto               1                   Bar   
33      West Toronto               1        Breakfast Spot   
34      West Toronto               1                  Café   
35      West Toronto               1        Breakfast Spot   
36      West Toronto               1           Coffee Shop   
37      East Toronto               1    Light Rail Station   

    2nd Most Common Venue 3rd Most Common Venue  \
1             Coffee Shop        Ice Cream Shop   
2             Coffee Shop        Ice Cream Shop   
3             Coffee Shop    Italian Restaurant   
5        Asian Restaurant                 Hotel   
6             Coffee Shop           Yoga Studio   
7             Pizza Place          Dessert Shop   
8                     Gym            Playground   
11             Restaurant                  Café   
12    Japanese Restaurant               Gay Bar   
13                 Bakery                  Café   
14         Clothing Store                  Café   
15             Restaurant                  Café   
16           Cocktail Bar            Restaurant   
17     Italian Restaurant        Sandwich Place   
18        Thai Restaurant                  Café   
19               Aquarium                 Hotel   
20                   Café                 Hotel   
21                  Hotel                  Café   
24         Sandwich Place                  Café   
25                Theater                Bakery   
26  Vietnamese Restaurant                   Bar   
27        Airport Service        Airport Lounge   
28             Restaurant                  Café   
29                   Café                 Hotel   
30          Grocery Store                  Park   
31                 Bakery        Discount Store   
32       Asian Restaurant           Coffee Shop   
33            Coffee Shop                  Café   
34                    Bar    Mexican Restaurant   
35              Gift Shop                   Bar   
36                   Café           Pizza Place   
37                 Garden         Garden Center   

            4th Most Common Venue      5th Most Common Venue  \
1              Italian Restaurant          Indian Restaurant   
2                    Liquor Store         Light Rail Station   
3                       Gastropub        American Restaurant   
5               Food & Dri

#### Toronto Festival Cluster

We have named this cluster as a festival cluster because this area is very suitable if someone who wants to arrange an event or a festival. There is an event space with numerous restaurants around it so catering will be very easy with electronic store to support any electronic needs for the businesses. 

In [232]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               2                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22   Filipino Restaurant        Farmers Market    Falafel Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22           Event Space  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

#### Toronto Picnic Spot Cluster

This cluster has been named as picnic spot cluster because there is park with a trail. There are also multiple restaurants around this cluster so it will be very convenient for the families to have an outdoor activity in this area. Pet lovers will also enjoy the spot because there is a Dog Run venue as well. In short it is very good out door venue for families and friends.

In [233]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10  Downtown Toronto               3                  Park   
23   Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10            Playground                 Trail           Wings Joint   
23      Sushi Restaurant                 Trail         Jewelry Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10   Dumpling Restaurant                 Diner        Discount Store   
23           Wings Joint            Donut Shop        Discount Store   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
10               Dog Run      Doner Restaurant                   Donut Shop  
23               Dog Run      Doner Restaurant  Eastern European Restaurant

#### Toronto Community center Cluster

This cluster looks like community center because there is bus line so the transport is very convenient. There is a park, restaurants and stores which makes this place useful of everyone in the community to go in this area and buy food or other useful items. 

In [234]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4  Central Toronto               4              Bus Line   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4                  Park           Swim School           Wings Joint   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4    Falafel Restaurant           Event Space  Ethiopian Restaurant   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
4     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

#### This is the End of 3rd requirement where we have to cluster the neighborhoods of toronto and Examine them. Thank You!!!